### PR TITLE
Add ENOBUFS handling for unsolicited messages

### PR DESCRIPTION
**Pull request by Tuetuopay**
This can happen when large burst of messages come all of a sudden, which happen very easily when routing protocols are involved (e.g. BGP). The current implementation incorrectly assumes that any failure to read from the socket is akin to the socket closed. This is not the case.

This adds handling for this specific error, which translates to a wrapper struct in the unsolicited messages stream: either a message, or an overrun. This lets applications handle best for their usecase such event: either resync because messages are lost, or do nothing if the listening is informational only (e.g. logging).

This is a direct port of https://github.com/little-dude/netlink/pull/293.

### CODE REVIEW

File: src/framed.rs [hunk 1]
```diff
@@ -63,6 +67,23 @@ where
...
+                // more message can fit in), new messages will be dropped and recv calls will
+                // return `ENOBUFS`.
+                // This needs to be handled for applications to resynchronize with the contents
+                // of the kernel if necessary.
+                // We don't need to do anything special:
+                // - contents of the reader is still valid because we won't have partial messages
+                //   in there anyways (large enough buffer)
+                // - contents of the socket's internal buffer is still valid because the kernel
+                //   won't put partial data in it
+                Err(e) if e.raw_os_error() == Some(105) => {
```
Comment (cathay4t) [inline src/framed.rs:82]: Instead of using magic number, please use

```
const ENOBUFS: i32 = 105;
```

**Review by cathay4t — CHANGES_REQUESTED** [review]

**Comment by cathay4t** [comment]
@Tuetuopay The kernel is treating is as an error. With the changes you suggested, almost every existing rust-netlink crate needs to be changed and their user also just for a rare use case. The rust has a saying: you do not need to pay for what you will not use. Clearly your patch is against this design.

My suggestion is, raise proper error like `OverRunError`, for any user who cares, they can handle this error by a  controlled retry.

File: src/framed.rs (hunk 1 above)
Comment (Tuetuopay) [reply src/framed.rs]: Indeed. Updated with the constant (and rebased on `main` at the same time).

**Comment by cathay4t** [comment]
I prefer `NetlinkPayload::Overrun` way.
Meanwhile, the existing user facing crate(e.g. rtnetlink) might need to amend to support this in their `macros.rs` to convert ``NetlinkPayload::Overrun` to `Err(rtnetlink::Error::Overrun)`.  So user of rtnetlink could ignore this specific error.
To even better, we can allow user to say `AddressGetRequest::set_auto_retry_on_overrun(true)` or `AddressGetRequest::set_ignore_overrun()` to make the user's life eaiser.

**Comment by Tuetuopay** [comment]
> I prefer `NetlinkPayload::Overrun` way.

Sure! Implementing this.

> Meanwhile, the existing user facing crate(e.g. rtnetlink) might need to amend to support this in their `macros.rs` to convert `NetlinkPayload::Overrun` to `Err(rtnetlink::Error::Overrun)`. So user of rtnetlink could ignore this specific error. To even better, we can allow user to say `AddressGetRequest::set_auto_retry_on_overrun(true)` or `AddressGetRequest::set_ignore_overrun(true)` to make the user's life eaiser.

Okay there is a deep misunderstanding of the issue here. This does not fixes issues with request/reply mode (as suggested by `AddressGetRequest::`), but with subscription mode. In this mode, the kernel sends notifications as fast as it can, without waiting. This does not happen in regular request/response mode because it will wait for consumers to ack messages before sending more.

Basically we can see regular request/response mode as TCP and multicast mode as UDP, if you squint hard enough.

As for allowing the user to ignore those errors, they can already be disabled when opening the stream socket with `Socket::set_no_enobufs`.

**Comment by Tuetuopay** [comment]
Hi @cathay4t!

I pushed a fully new implementation. This is what was suggested: generate actual `NetlinkMessage`s with a payload of `NetlinkPayload::Overrun`. This way, users can continue ignoring the error while people that mind (e.g. BGP people) can properly handle the error.

On a related note, I saw https://github.com/rust-netlink/netlink-packet-core/pull/7 and it can indeed be done as I do not use the struct anymore in the new version.

I hope this one is more to your taste :)

**Comment by cathay4t** [comment]
@Tuetuopay Yes. The new patch looks good. Please rebase it then we are OK to merge.

Do you have plan to change rtnetlink `try_rtnl!` for this?

**Comment by cathay4t** [comment]
The `src/connection.rs` has `Overrun(_) => unimplemented!("overrun is not handled yet")` , can you also patch it?

**Comment by yshui** [comment]
I am also hitting the ENOBUFS problem. Did @Tuetuopay lost interests in this PR? I can try to take over as this doesn't look too complex.

**Comment by Tuetuopay** [comment]
> I am also hitting the ENOBUFS problem. Did @Tuetuopay lost interests in this PR? I can try to take over as this doesn't look too complex.

Pretty much, I stopped using it for the usecase I was hitting. I may use this library again in the future though.

I'll see to address the comments, but if I don't get time for this feel free to pick it up :)

**Comment by Tuetuopay** [comment]
> The `src/connection.rs` has `Overrun(_) => unimplemented!("overrun is not handled yet")` , can you also patch it?

@cathay4t Well, about that, it was left as unimplemented on purpose. You will get ENOBUFS on the _unsolicited_ message channel, not on the regular request/response channel. Why? Because when using netlink in request/response mode, there is an ack mechanism between the kernel and userspace, precisely to avoid this issue. (well, unless the buffer sizes are set stupidly low, but in such case, this is a non-recoverable error).

I can set it as a fatal error, and make it a `break` from the `forward_responses` loop to exit the whole netlink system. Less violent than the `unimplemented!`. Your call :)

**Comment by Tuetuopay** [comment]
> @Tuetuopay Yes. The new patch looks good. Please rebase it then we are OK to merge.
> 
> Do you have plan to change rtnetlink `try_rtnl!` for this?

Just remembered this. Well, no, for the same reason I left the unimplemented: this is not supposed to happen in a request/response scenario. It _really_ is an unexpected thing to get in req/res.

**Review by cathay4t — APPROVED** [review]

**Comment by cathay4t** [comment]
Sorry for the long-long wait. The patch has shipped to version 0.11.4.
